packages/outbox-core/lib/outbox.ts

```diff
@@ -0,0 +1,139 @@
+import { AbstractPeriodicJob, type JobExecutionContext } from '@lokalise/background-jobs-common'
+import type { PeriodicJobDependencies } from '@lokalise/background-jobs-common'
+import type {
+  CommonEventDefinition,
+  CommonEventDefinitionPublisherSchemaType,
+  ConsumerMessageMetadataType,
+  DomainEventEmitter,
+} from '@message-queue-toolkit/core'
+import { PromisePool } from '@supercharge/promise-pool'
+import { uuidv7 } from 'uuidv7'
+import type { OutboxAccumulator } from './accumulators'
+import type { OutboxStorage } from './storage'
+
+export type OutboxDependencies<SupportedEvents extends CommonEventDefinition[]> = {
+  outboxStorage: OutboxStorage<SupportedEvents>
+  outboxAccumulator: OutboxAccumulator<SupportedEvents>
+  eventEmitter: DomainEventEmitter<SupportedEvents>
+}
+
+export type OutboxProcessorConfiguration = {
+  maxRetryCount: number
+  emitBatchSize: number
+}
+
+export type OutboxConfiguration = {
+  jobIntervalInMs: number
+} & OutboxProcessorConfiguration
+
+/**
+ * Main logic for handling outbox entries.
+ *
+ * If entry is rejected, it is NOT going to be handled during the same execution. Next execution will pick it up.
+ */
+export class OutboxProcessor<SupportedEvents extends CommonEventDefinition[]> {
+  constructor(
+    private readonly outboxDependencies: OutboxDependencies<SupportedEvents>,
+    private readonly outboxProcessorConfiguration: OutboxProcessorConfiguration,
+  ) {}
+
+  public async processOutboxEntries(context: JobExecutionContext) {
+    const { outboxStorage, eventEmitter, outboxAccumulator } = this.outboxDependencies
+
+    const entries = await outboxStorage.getEntries(this.outboxProcessorConfiguration.maxRetryCount)
+
+    const currentEntriesInAccumulator = new Set(
```

Comment:
we can skip this step and the next one altogether if `entries.length === 0`, which is going to be the most typical case

Reply:
Done,